internal/pkg/poster/poster.go

```diff
@@ -70,154 +80,159 @@ func (p *Poster) StopPoSting() {
 }
 
 func (p *Poster) startPoStIfNeeded(ctx context.Context, newHead block.TipSet) error {
-	// TODO: Window PoSt (#4010)
-	//p.postMutex.Lock()
-	//defer p.postMutex.Unlock()
-	//
-	//if p.postCancel != nil {
-	//	// already posting
-	//	return nil
-	//}
-	//
-	//tipsetHeight, err := newHead.Height()
-	//if err != nil {
-	//	return err
-	//}
-	//
-	//root, err := p.chain.GetTipSetStateRoot(ctx, newHead.Key())
-	//if err != nil {
-	//	return err
-	//}
-	//
-	//stateView := p.stateViewer.StateView(root)
-	//provingPeriodStart, _, _, err := stateView.MinerProvingPeriod(ctx, p.minerAddr)
-	//if err != nil {
-	//	return nil
-	//}
-	//
-	//if provingPeriodStart > tipsetHeight {
-	//	// it's not time to PoSt
-	//	return nil
-	//}
-	//
-	//ctx, p.postCancel = context.WithCancel(ctx)
-	//go p.doPoSt(ctx, stateView, provingPeriodStart, newHead.Key())
-	//
+	p.postMutex.Lock()
+	defer p.postMutex.Unlock()
+
+	if p.postCancel != nil {
+		// already posting
+		return nil
+	}
+
+	tipsetHeight, err := newHead.Height()
+	if err != nil {
+		return err
+	}
+
+	root, err := p.chain.GetTipSetStateRoot(ctx, newHead.Key())
+	if err != nil {
+		return err
+	}
+
+	stateView := p.stateViewer.StateView(root)
+	index, open, _, challengeAt, err := stateView.MinerDeadlineInfo(ctx, p.minerAddr, tipsetHeight)
+	if err != nil {
+		return err
+	}
+
+	// exit if we haven't yet hit the deadline
+	if tipsetHeight < open {
+		return nil
+	}
+
+	randomness, err := p.getChallenge(ctx, newHead.Key(), challengeAt)
+	if err != nil {
+		return err
+	}
+
+	// If we have already seen this randomness, either the deadline has changed or
+	// or the chain as reorged to a point prior to the challenge. Either way,
+	// it is time to start a new PoSt.
+	if bytes.Equal(p.challenge, randomness) {
```

Comment:
I'm not sure I see how the comment and code line up.  It seems like "we have already seen this randomness" would correspond to this if statement condition, but in that case we *do not* start a new PoSt.

Reply:
good call. I added a "not".